frontends/concrete-python/tests/compilation/test_modules.py

```diff
@@ -429,3 +418,151 @@ def dec(x):
+    assert isinstance(Module.composition, NotComposable)
+
+
+def test_composition_policy_all_composable():
+    @fhe.module()
+    class Module:
+        @fhe.function({"x": "encrypted"})
+        def square(x):
+            return x**2
+
+        @fhe.function({"x": "encrypted", "y": "encrypted"})
+        def add_sub(x, y):
+            return (x + y), (x - y)
+
+        @fhe.function({"x": "encrypted", "y": "encrypted"})
+        def mul(x, y):
+            return x * y
+
+        composition = ModuleComposable()
+
+    assert isinstance(Module.composition, CompositionPolicy)
+    assert isinstance(Module.composition, ModuleComposable)
+
+
+def test_composition_policy_wires():
+    @fhe.module()
+    class Module:
+        @fhe.function({"x": "encrypted"})
+        def square(x):
+            return x**2
+
+        @fhe.function({"x": "encrypted", "y": "encrypted"})
+        def add_sub(x, y):
+            return (x + y), (x - y)
+
+        composition = Wired(
+            [
+                Wire(AllOutputs(add_sub), AllInputs(add_sub)),
+                Wire(AllOutputs(add_sub), Input(square, 0)),
+            ]
+        )
+
+    assert isinstance(Module.composition, CompositionPolicy)
+    assert isinstance(Module.composition, Wired)
+
+
+def test_composition_wired_enhances_complexity():
+    @fhe.module()
+    class Module1:
+        @fhe.function({"x": "encrypted"})
+        def _1(x):
+            return (x * 2) % 20
+
+        @fhe.function({"x": "encrypted"})
+        def _2(x):
+            return (x * 2) % 200
+
+        composition = Wired(
+            [
+                Wire(Output(_1, 0), Input(_2, 0)),
+            ]
+        )
+
+    module1 = Module1.compile(
+        {
+            "_1": [np.random.randint(1, 20, size=()) for _ in range(100)],
+            "_2": [np.random.randint(1, 200, size=()) for _ in range(100)],
+        },
+    )
+
+    @fhe.module()
+    class Module2:
+        @fhe.function({"x": "encrypted"})
+        def _1(x):
+            return (x * 2) % 20
+
+        @fhe.function({"x": "encrypted"})
+        def _2(x):
+            return (x * 2) % 200
+
+        composition = ModuleComposable()
+
+    module2 = Module2.compile(
+        {
+            "_1": [np.random.randint(1, 20, size=()) for _ in range(100)],
+            "_2": [np.random.randint(1, 200, size=()) for _ in range(100)],
+        },
+    )
+
+    assert module1.complexity < module2.complexity
+
+
+def test_composition_wired_compilation():
+    @fhe.module()
+    class Module:
+        @fhe.function({"x": "encrypted"})
+        def a(x):
+            return (x * 2) % 20
+
+        @fhe.function({"x": "encrypted"})
+        def b(x):
+            return (x * 2) % 50
+
+        @fhe.function({"x": "encrypted"})
+        def c(x):
+            return (x * 2) % 100
+
+        composition = Wired(
+            [
+                Wire(Output(a, 0), Input(b, 0)),
+                Wire(Output(b, 0), Input(c, 0)),
+            ]
+        )
```

Comment:
What happened to this API:
```python
    wires = [
        (square[0], add_sub.x),
        (square[0], add_sub.y),
        (add_sub[0], mul.x),
        (add_sub[1], mul.y),
    ]
```

Since we have a custom annotation (fhe.function), it should be doable. It's much more consise IMO. Especially with named inputs.

Reply:
(it's okay to keep the internal structure, this would just be a wrapper)

Reply:
The current API allows more diverse patterns than what the original API allowed, like with `ModuleComposable` or `AllOutputs` `AllInputs`. It is true that it is less concise, but it makes things clearer in my opinion. It is likely a matter of taste !
I don't mind changing the api at all, but maybe we should ask ml guys which one they prefer ? Also if they prefer the original one, we'll have to find a way to write `ModuleComposable` and `AllInputs/AllOutputs`.

Reply:
We can provide both APIs if needed. But IMO, this is the best API:
- Specific composition
```
    wires = [
        (square[0], add_sub.x),
        (square[0], add_sub.y),
        (add_sub[0], mul.x),
        (add_sub[1], mul.y),
    ]
```
- To all inputs
```
    wires = [
        (square[0], add_sub),
        (square[0], add_sub),
        (add_sub[0], mul),
        (add_sub[1], mul),
    ]
```
- From all outputs
```
    wires = [
        (square, add_sub.x),
        (square, add_sub.y),
        (add_sub, mul.x),
        (add_sub, mul.y),
    ]
```
- From all outputs to all inputs
```
    wires = [
        (square, add_sub),
        (add_sub, mul),
    ]
```
- Everything
```
composable=True
```

Reply:
Given the output of ml, I propose we keep it the way it is for now, and if ml prefers a different API, I'll implement it. In any case, I'll let the current API as inner abstraction, so this would not change anything

Reply:
Sure, the additional API can be provided in a different PR :+1: